src/EliminateBoolVectors.cpp

```diff
@@ -0,0 +1,146 @@
+#include "IRMutator.h"
+#include "IROperator.h"
+
+namespace Halide {
+namespace Internal {
+
+class EliminateBoolVectors : public IRMutator {
+private:
+    using IRMutator::visit;
+
+    template <typename T>
+    void visit_comparison(const T* op) {
+        Expr a = mutate(op->a);
+        Expr b = mutate(op->b);
+        Type t = a.type();
+
+        // Ensure both a and b have the same type (if this is a vector
+        // comparison). This should only be necessary if the operands are
+        // integer vectors (promoted from bool vectors).
+        if (t.lanes() > 1 && t.bits() != b.type().bits()) {
+            internal_assert(t.is_int() && b.type().is_int());
+
+            t = t.with_bits(std::max(t.bits(), b.type().bits()));
+            if (t != a.type()) {
+                a = Cast::make(t, a);
+            }
+            if (t != b.type()) {
+                b = Cast::make(t, b);
+            }
+        }
+
+        if (!a.same_as(op->a) || !b.same_as(op->b)) {
+            expr = T::make(a, b);
+        } else {
+            expr = op;
+        }
+
+        if (t.lanes() > 1) {
+            // To represent bool vectors, OpenCL uses vectors of signed
+            // integers with the same width as the types being compared.
+            t = t.with_code(Type::Int);
+            expr = Cast::make(t, expr);
```

Comment:
If we need to modify this to always generate booleans of a particular size, we can do that. Though, I think it's correct as far as I can tell. vmux looks like it takes registers of the same size as the other operands (at least in 64 byte mode), so it should work. It's not clear from the docs if the predicate register is a different size in 128 byte mode or not.